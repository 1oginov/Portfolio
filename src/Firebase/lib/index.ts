import * as firebase from 'firebase/app';
import 'firebase/auth';

import * as C from 'Constants';

export interface GetFirebaseExtraArgument {
  (): firebase.app.App;
}

export interface WithFirebaseHocProps {
  firebase: firebase.app.App;
}

export const getAuthProvider = (provider: string): firebase.auth.AuthProvider => {
  switch (provider) {
    case 'google':
      return new firebase.auth.GoogleAuthProvider();

    default:
      throw new Error('Unknown auth provider');
  }
};

export const getPositionsCollectionPath = (userId: string): string => (
  `${C.FIRESTORE_USERS_COLLECTION}/${userId}/${C.FIRESTORE_POSITIONS_COLLECTION}`
);

export const getPositionDocumentPath = (userId: string, positionId: string): string => (
  `${C.FIRESTORE_USERS_COLLECTION}/${userId}/${C.FIRESTORE_POSITIONS_COLLECTION}/${positionId}`
);

export const getPositionsCollection = (
  firebaseInstance: firebase.app.App,
  userId: string,
): firebase.firestore.CollectionReference => (
  firebaseInstance.firestore()
    .collection(C.FIRESTORE_USERS_COLLECTION)
    .doc(userId)
    .collection(C.FIRESTORE_POSITIONS_COLLECTION)
);

export const getPositionDocument = (
  firebaseInstance: firebase.app.App,
  userId: string,
  positionId: string,
): firebase.firestore.DocumentReference => getPositionsCollection(firebaseInstance, userId).doc(positionId);
