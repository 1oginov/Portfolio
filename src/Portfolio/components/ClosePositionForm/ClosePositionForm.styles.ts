import { withStyles } from '@material-ui/core';

export default withStyles(theme => ({
  root: {
    display: 'flex',
    flexDirection: 'column',
    width: '100%',
  },
  row: {
    margin: theme.spacing(1),
    textAlign: 'center',
  },
}));
