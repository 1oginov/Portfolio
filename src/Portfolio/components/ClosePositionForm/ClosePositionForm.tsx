import * as React from 'react';

export interface Props {
  commission: number;
  date: string;
  handleCommissionChange: (event: React.SyntheticEvent) => void;
  handleDateChange: (event: React.SyntheticEvent) => void;
  handlePriceChange: (event: React.SyntheticEvent) => void;
  handleSubmit: (event: React.SyntheticEvent) => void;
  price: number;
}

const ClosePositionForm: React.FunctionComponent<Props> = ({
  commission, date, handleCommissionChange, handleDateChange, handlePriceChange, handleSubmit, price,
}: Props) => (
  <form onSubmit={handleSubmit}>

    <div>
      Price
      <br />
      <input
        min="0"
        onChange={handlePriceChange}
        step="0.01"
        type="number"
        value={price}
      />
    </div>

    <div>
      Commission
      <br />
      <input
        min="0"
        onChange={handleCommissionChange}
        step="0.01"
        type="number"
        value={commission}
      />
    </div>

    <div>
      Date
      <br />
      <input onChange={handleDateChange} type="date" value={date} />
    </div>

    <div>
      <button type="submit">Close</button>
    </div>

  </form>
);

export default ClosePositionForm;
