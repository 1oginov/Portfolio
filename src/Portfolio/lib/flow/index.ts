/* @flow */

export type Position = {
  id: string,
  symbol: string,
  price: number,
  amount: number,
  date: string,
};
