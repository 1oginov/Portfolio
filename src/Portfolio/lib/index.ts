// TODO: Tests.

export interface Position {
  id: string;
  symbol: string;
  amount: number;

  openPrice: number;
  openCommission: number;
  openDate: string;

  closePrice: number | null;
  closeCommission: number | null;
  closeDate: string | null;
}

export const formatDate = (date: Date): string => date.toISOString().slice(0, 10);

export const getUniqueSymbolsFromPositions = (positions: Position[]): string[] => Array.from(
  new Set(positions.map(position => position.symbol)),
);
