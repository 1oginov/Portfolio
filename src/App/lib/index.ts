import { annualizePercent, Position } from 'Portfolio/lib';
import { getQuotePrice } from 'Stocks/lib';
import Quote from 'Stocks/lib/IEX/Quote';

// TODO: Tests.

export interface StockPosition extends Position {
  openSum: number;

  closeSum: number | null;
  closePL: number | null;
  closePLPercent: number | null;
  closePLAnnualPercent: number | null;

  quote: Quote | null;
  quoteProgress: boolean;

  companyName: string | null;
  dailyPL: number | null;
  dailyPLPercent: number | null;

  marketPrice: number | null;
  marketSum: number | null;
  marketPL: number | null;
  marketPLPercent: number | null;
  marketPLAnnualPercent: number | null;
}

export interface OpenPositionsSummary {
  symbol: string;
  amount: number;

  openPriceAverage: number;
  openSum: number;

  companyName: string | null;
  dailyPL: number | null;
  dailyPLPercent: number | null;

  marketPL: number | null;
  marketPLPercent: number | null;
}

export interface Totals {
  totalOpenSum: number;

  totalCloseSum: number;
  totalClosePL: number;
  totalClosePLPercent: number;

  totalDailyPL: number;
  totalDailyPLPercent: number;

  totalMarketPL: number;
  totalMarketPLPercent: number;
  totalMarketSum: number;
}

export const createStockPosition = (position: Position, quote: Quote | null, quoteProgress: boolean): StockPosition => {
  const openSum = position.amount * position.openPrice + position.openCommission;

  let closeSum = null;
  let closePL = null;
  let closePLPercent = null;
  let closePLAnnualPercent = null;

  if (position.closeCommission !== null && position.closeDate !== null && position.closePrice !== null) {
    closeSum = position.amount * position.closePrice - position.closeCommission;
    closePL = closeSum - openSum;
    closePLPercent = closePL / openSum;
    closePLAnnualPercent = annualizePercent(closePLPercent, position.openDate, position.closeDate);
  }

  let companyName = null;
  let dailyPL = null;
  let dailyPLPercent = null;

  if (quote !== null) {
    ({ companyName } = quote);
    dailyPL = position.amount * quote.change;
    dailyPLPercent = quote.changePercent;
  }

  const marketPrice = getQuotePrice(quote);
  let marketSum = null;
  let marketPL = null;
  let marketPLPercent = null;
  let marketPLAnnualPercent = null;

  if (marketPrice !== null) {
    marketSum = position.amount * marketPrice;
    marketPL = marketSum - openSum;
    marketPLPercent = marketPL / openSum;
    marketPLAnnualPercent = annualizePercent(marketPLPercent, position.openDate);
  }

  return {
    ...position,
    closePL,
    closePLAnnualPercent,
    closePLPercent,
    closeSum,
    companyName,
    dailyPL,
    dailyPLPercent,
    marketPL,
    marketPLAnnualPercent,
    marketPLPercent,
    marketPrice,
    marketSum,
    openSum,
    quote,
    quoteProgress,
  };
};

export const createOpenPositionsSummaries = (stockPositions: StockPosition[]): OpenPositionsSummary[] => {
  const bySymbols: { [key: string]: StockPosition[] } = {};

  stockPositions.forEach(position => {
    if (!bySymbols[position.symbol]) {
      bySymbols[position.symbol] = [];
    }

    bySymbols[position.symbol].push(position);
  });

  return Object.keys(bySymbols).map(symbol => {
    let amount = 0;
    let openPrice = 0;
    let openSum = 0;
    let dailyPL: number | null = null;
    let dailyPLPercent: number | null = null;
    let marketPL: number | null = null;
    let marketPLPercent: number | null = null;

    bySymbols[symbol].forEach(stockPosition => {
      amount += stockPosition.amount;
      openPrice += stockPosition.amount * stockPosition.openPrice;
      openSum += stockPosition.openSum;

      if (stockPosition.dailyPL !== null) {
        if (dailyPL === null) {
          dailyPL = 0;
        }

        dailyPL += stockPosition.dailyPL;
      }

      if (stockPosition.marketPL !== null) {
        if (marketPL === null) {
          marketPL = 0;
        }

        marketPL += stockPosition.marketPL;
      }
    });

    const referencePosition = bySymbols[symbol][0];

    if (dailyPL !== null && referencePosition.quote !== null) {
      dailyPLPercent = dailyPL / amount / referencePosition.quote.close;
    }

    if (marketPL !== null) {
      marketPLPercent = marketPL / openSum;
    }

    const openPriceAverage = openPrice / amount;
    const { companyName } = referencePosition;

    return {
      amount,
      companyName,
      dailyPL,
      dailyPLPercent,
      marketPL,
      marketPLPercent,
      openPriceAverage,
      openSum,
      symbol,
    };
  });
};

export const calculateTotals = (stockPositions: StockPosition[]): Totals => {
  let totalOpenSum = 0;
  let totalCloseSum = 0;
  let totalDailyPL = 0;
  let totalPreviousCloseSum = 0;
  let totalMarketSum = 0;

  stockPositions.forEach(position => {
    totalOpenSum += position.openSum;

    if (position.closeSum !== null) {
      totalCloseSum += position.closeSum;
    }

    if (position.quote !== null) {
      totalDailyPL += position.amount * position.quote.change;
      totalPreviousCloseSum += position.amount * position.quote.close;
    }

    if (position.marketSum !== null) {
      totalMarketSum += position.marketSum;
    }
  });

  const totalClosePL = totalCloseSum - totalOpenSum;
  const totalClosePLPercent = totalClosePL / totalOpenSum;

  const totalDailyPLPercent = totalDailyPL / totalPreviousCloseSum;

  const totalMarketPL = totalMarketSum - totalOpenSum;
  const totalMarketPLPercent = totalMarketPL / totalOpenSum;

  return {
    totalClosePL,
    totalClosePLPercent,
    totalCloseSum,
    totalDailyPL,
    totalDailyPLPercent,
    totalMarketPL,
    totalMarketPLPercent,
    totalMarketSum,
    totalOpenSum,
  };
};
