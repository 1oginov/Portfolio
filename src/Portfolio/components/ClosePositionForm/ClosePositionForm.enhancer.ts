import { connect } from 'react-redux';
import { compose, withHandlers, withStateHandlers } from 'recompose';

import { Props } from './ClosePositionForm';
import { closePosition as closePositionAction, ClosePositionAction } from '../../actions';
import { formatDate } from '../../lib';

const mapDispatchToProps = { closePosition: closePositionAction };

interface DispatchProps {
  closePosition: ClosePositionAction;
}

interface WithStateHandlersProps {
  commission: number;
  date: string;
  price: number;
}

interface EnhancedProps {
  id: string;
  onClose?: (positionId: string) => void;
}

export default compose<Props & DispatchProps & WithStateHandlersProps, EnhancedProps>(
  connect(null, mapDispatchToProps),
  withStateHandlers(
    {
      commission: 0,
      date: formatDate(new Date()),
      price: 0,
    },
    {
      handleCommissionChange: () => event => {
        const price = parseFloat(event.target.value);

        return {
          commission: price >= 0 ? price : 0,
        };
      },
      handleDateChange: () => event => {
        const date = new Date(event.target.value);

        return {
          date: formatDate(date || new Date()),
        };
      },
      handlePriceChange: () => event => {
        const price = parseFloat(event.target.value);

        return {
          price: price >= 0 ? price : 0,
        };
      },
    },
  ),
  withHandlers<EnhancedProps & DispatchProps & WithStateHandlersProps, {}>({

    handleSubmit: ({
      closePosition, commission, date, id, onClose, price,
    }) => (event: React.SyntheticEvent) => {
      event.preventDefault();

      closePosition(id, price, commission, date)
        .then(positionId => {
          if (onClose) {
            onClose(positionId);
          }
        });
    },

  }),
);
