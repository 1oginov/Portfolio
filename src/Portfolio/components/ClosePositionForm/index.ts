import ClosePositionForm from './ClosePositionForm';
import enhance from './ClosePositionForm.enhancer';

export default enhance(ClosePositionForm);
