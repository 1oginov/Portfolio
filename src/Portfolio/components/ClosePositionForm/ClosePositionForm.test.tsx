/* eslint-disable import/no-extraneous-dependencies */

import { shallow } from 'enzyme';
import classes from 'identity-obj-proxy';
import * as React from 'react';

import { formatDate } from 'Shared/lib';

import ClosePositionForm from './ClosePositionForm';

it('matches snapshot', () => {
  const wrapper = shallow(
    <ClosePositionForm
      classes={classes}
      commission={1}
      date={formatDate(new Date('2019-07-02'))}
      handleCommissionChange={() => undefined}
      handleDateChange={() => undefined}
      handlePriceChange={() => undefined}
      handleSubmit={() => undefined}
      price={100}
    />,
  );

  expect(wrapper).toMatchSnapshot();
});
