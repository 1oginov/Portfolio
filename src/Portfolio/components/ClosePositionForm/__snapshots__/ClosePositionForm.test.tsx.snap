// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<form
  onSubmit={[Function]}
>
  <div>
    Price
    <br />
    <input
      min="0"
      onChange={[Function]}
      step="0.01"
      type="number"
      value={100}
    />
  </div>
  <div>
    Commission
    <br />
    <input
      min="0"
      onChange={[Function]}
      step="0.01"
      type="number"
      value={1}
    />
  </div>
  <div>
    Date
    <br />
    <input
      onChange={[Function]}
      type="date"
      value="2019-07-02"
    />
  </div>
  <div>
    <button
      type="submit"
    >
      Close
    </button>
  </div>
</form>
`;
